lib/os/heap-validate.c

```diff
@@ -0,0 +1,307 @@
+/*
+ * Copyright (c) 2019 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#include <sys/sys_heap.h>
+#include <kernel.h>
+#include "heap.h"
+
+/* White-box sys_heap validation code.  Uses internal data structures.
+ * Not expected to be useful in production apps.  This checks every
+ * header field of every chunk and returns true if the totality of the
+ * data structure is a valid heap.  It doesn't necessarily tell you
+ * that it is the CORRECT heap given the history of alloc/free calls
+ * that it can't inspect.  In a pathological case, you can imagine
+ * something scribbling a copy of a previously-valid heap on top of a
+ * running one and corrupting it. YMMV.
+ */
+
+static size_t max_chunkid(struct z_heap *h)
+{
+	return h->len - bytes_to_chunksz(h, 1);
+}
+
+static bool in_bounds(struct z_heap *h, chunkid_t c)
+{
+	return (c >= h->chunk0)
+		&& (c <= max_chunkid(h))
+		&& (size(h, c) < h->len);
+}
+
+static bool valid_chunk(struct z_heap *h, chunkid_t c)
+{
+	return (size(h, c) > 0
+		&& (c + size(h, c) <= h->len)
+		&& in_bounds(h, c)
+		&& ((c == h->chunk0) || in_bounds(h, c - left_size(h, c)))
+		&& (used(h, c) || in_bounds(h, free_prev(h, c)))
+		&& (used(h, c) || in_bounds(h, free_next(h, c))));
+}
+
+/* Validate multiple state dimensions for the bucket "next" pointer
+ * and see that they match.  Probably should unify the design a
+ * bit...
+ */
+static inline void check_nexts(struct z_heap *h, int bidx)
+{
+	struct z_heap_bucket *b = &h->buckets[bidx];
+
+	bool emptybit = (h->avail_buckets & (1 << bidx)) == 0;
+	bool emptylist = b->next == 0;
+	bool emptycount = b->list_size == 0;
+	bool empties_match = emptybit == emptylist && emptybit == emptycount;
```

Comment:
Parenthesis are not needed here, but very good to have. They will speak your true intentions over guessed ones in the future.